File: 1005.k-次取反后最大化的数组和.c
/*
 * @lc app=leetcode.cn id=1005 lang=c
 *
 * [1005] K 次取反后最大化的数组和
 */

// @lc code=start

int largestSumAfterKNegations(int *nums, int numsSize, int k)
{
  // 要按绝对值大小排序
  int i, j, sum, temp;
  sum = 0;
  for (i = 0; i < numsSize - 1; i++)
  {
    for (j = 0; j < numsSize - i - 1; j++)
    {
      if (nums[j + 1] < nums[j])
      {
        temp = nums[j];
        nums[j] = nums[j + 1];
        nums[j + 1] = temp;
      }
    }
  }
  for (i = 0; i < numsSize; i++)
  {
    if (nums[i] <= 0 && k > 0)
    {
      nums[i] = -nums[i];
      k--;
    }
  }
  for (i = 0; i < numsSize - 1; i++)
  {
    for (j = 0; j < numsSize - i - 1; j++)
    {
      if (nums[j + 1] < nums[j])
      {
        temp = nums[j];
        nums[j] = nums[j + 1];
        nums[j + 1] = temp;
      }
    }
  }
  while (k > 0)
  {
    nums[0] *= -1;
    k--;
    if (k == 0)
      break;
  }

  for (i = 0; i < numsSize; i++)
    sum += nums[i];

  return sum;
}
// @lc code=end

File: store/1005.k-次取反后最大化的数组和.js
/*
 * @lc app=leetcode.cn id=1005 lang=javascript
 *
 * [1005] K 次取反后最大化的数组和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const absSort = (arr) => {
  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr.length - i - 1; j++) {
      if (Math.abs(arr[j]) < Math.abs(arr[j + 1])) {
        let temp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = temp;
      }
    }
  }
};
var largestSumAfterKNegations = function (nums, k) {
  absSort(nums);
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] < 0) {
      nums[i] *= -1;
      k--;
    }
    if (!k) break;
  }
  if (k) for (let i = 0; i < k; i++) nums[nums.length - 1] *= -1;
  return nums.reduce((pre, cur) => pre + cur);
};
// @lc code=end
